### DISCUSSION TITLE
The necessity of fullscan

### DISCUSSION THREAD

**Discussion by JC-comp** [abraunegg/onedrive#2510] [Q&A] [answered]
Hi there,

Best to my research, the sync is rely on the /delta endpoint to track changes over time, and the changes will be stored in the database, that is we should always have a consistent snapshot of the online onedrive that can be used to sync with local files.
So, I am wondering what is the purpose of performing a fullscan to redo all the already seen changes.
Can we safely disable this action or under what scenario we cannot rely on the changes from /delta endpoint? That is not to do any fullscan "after we have a deltaUrl in the DB" under the following scenario with sync_list being used in monitor mode,
1. reach the monitor_fullscan_frequency
2. the fullscan for first sync loop 
3. the fullscan of True-Up in first sync loop 

I am asking for help because each fullscan took me over 14 hours to complete (over 1700000 OneDrive items processed, and I only want to sync a relative small part (one folder in root) of the drive, this took me a lot of time when I reboot or hit the monitor_fullscan_frequency interval.
Thanks in advance!

**Comment by abraunegg** [comment, marked as answer]
@JC-comp 
> over 1700000 OneDrive items processed

Firstly, you are 1.4 million items over what Microsoft advises you to store online. Please read https://support.microsoft.com/en-gb/office/restrictions-and-limitations-in-onedrive-and-sharepoint-64883a5d-228e-48f5-b3d2-eb39e07630fa#numberitemscanbesynced for details.

The Microsoft recommended maximum is 300,000 items before performance issues arise. My first advice to you is you need to drop your online folder size dramatically first and foremost if possible.

> So, I am wondering what is the purpose of performing a fullscan to redo all the already seen changes.

This is due to how the application was originally developed and the bugs / issues that the original source had (and still does) - the performing a periodic full scan was a way the original application handled those issues, which was carried over to this client. Most (if not all) of these bugs are fixed in the v2.4.x version, however v2.5.x is a 100% re-write to ensure that the end user experience is significantly improved.

> I only want to sync a relative small part (one folder in root) of the drive, this took me a lot of time when I reboot or hit the monitor_fullscan_frequency interval.

My suggestion here for you is the following:
* Dive deep into testing OneDrive Client Version v2.5.0-alpha-2 - read https://github.com/abraunegg/onedrive/discussions/2506 for details
* Configure this version as per the following:
  * Configure 'sync_list' to just sync the small part of what you want
  * Use `--sync --verbose --dry-run` and see where this gets you. 
  * If this is working as intended, test `--monitor --verbose --dry-run` for the same scenario
* If this is working all OK, please then test the following:
  * Remove 'sync_list' file
  * Use `--sync --single-directory 'path/to/folder/' --verbose --dry-run`
  * If this is working as intended, test `--monitor --single-directory 'path/to/folder/' --verbose --dry-run` for the same scenario

In both these scenario's the logic and functionality has been 100% re-written.

When using `--monitor` however, the same 300 second loop between /delta checks exists, so tweak 'monitor_interval' to be higher, but also adjust 'monitor_fullscan_frequency' to once per day perhaps or even higher (once per week?). Refer to the updated documentation here: https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-alpha-2/docs/application-config-options.md

Key point here to remember is: 'monitor_interval' x 'monitor_fullscan_frequency' = 'when full scan will be done'. Adjust these as necessary.

This being said, *all bets are off* with your environment if you are in any way using any Microsoft Azure Authentication Endpoint to comply with geographic data and security requirements. If you are having to use 'azure_ad_endpoint' - then in all likelihood /delta is not going to work for you at all, and v2.5.0-alpha-2 will have to generate a specific self-generated /delta compatible response.

If this is the case - the `--single-directory` path will be the most performant option for you, as it will *only* query that specific branch of your online data to self build a /delta response. This is a massive change to how v2.4.x and below operated.

**Reply by JC-comp** [reply]
Thanks for a super quick respose
Basically it is not quite possible for me to drop the size online for now, so under this constraint for the `onedrive-v2.5.0-alpha-2`, with `sync_list`  and `monitor` mode, the fullScan will only be triggered when there is no existing deltaLink or hit the `monitor_fullscan_frequency`, which is very close to what I expected. 
Without assigning `azure_ad_endpoint` you mentioned, I think I don't have to use`--single-directory` right? 
Because it seems that `--single-directory` mode will actually do a fullScan under the target directory for each `monitor_interval` (not `monitor_fullscan_frequency`), even if the number of items under the target folder do not exceed 300,000 items, it might still take a lot of time to sync in each loop.
So if using `sync_list` is enough, is there any document recording all the possible bugs or conditions that will cause inconsistency without full scan periodically? As this is the only case that will trigger the full scan in the new version, or acutally it is safe to stop this `monitor_fullscan_frequency`?

**Reply by abraunegg** [reply]
> Without assigning azure_ad_endpoint you mentioned, I think I don't have to use--single-directory right?

Not quite - my advice is try 'sync_list' and compare that to `--single-directory`. Both modes have been totally re-written to avoid major pulling down of data. When using  `--single-directory` it will build a self-compiled /delta response - which, essentially is the same net effect as if you were configuring `azure_ad_endpoint` - but, if you were using `azure_ad_endpoint` + 'sync_list' - the scan would be way more intensive.

> So if using sync_list is enough, is there any document recording all the possible bugs or conditions that will cause inconsistency without full scan periodically?

No such document exists at the moment, except when the application tells you that there is a sync issue, or something failed to sync, this is when that inconsistency will occur.

> As this is the only case that will trigger the full scan in the new version, or acutally it is safe to stop this monitor_fullscan_frequency?

There is no way to bypass the full scan at present. The only option here is that you set that value to some high number, so that in the normal course of operation - your sync loop counter is not hitting that value (thus triggering it). If you have normal system reboots and/or restart of the services - the net effect is that you should not hit performing the full scan at all when using v2.5.x

**Comment by abraunegg** [comment]
Resolved in 'alpha-2' where setting full scan to 0 will disable
